keps/753-descheduling-framework/README.md

```diff
@@ -0,0 +1,1211 @@
+
+
+
+### Descheduler
+
+The descheduler itself is expected to initialize the framework and invoke the high level framework routines responsible for executing individual plugin extension points. E.g.
+
+```go
+func New(...) (*Descheduler, error) {
+	...
+}
+
+func (d *Descheduler) descheduleOnce(ctx context.Context) error {
+	...
+	nodes, err := nodeutil.ReadyNodes(ctx, ...)
+	if status := d.framework.RunDeschedulePlugins(ctx, nodes); status != nil && status.Err != nil {
+		return status.Err
+	}
+
+	if status := d.framework.RunBalancePlugins(ctx, nodes); status != nil && status.Err != nil {
+		return status.Err
+	}
+
+	return nil
+}
+
+func (d *Descheduler) Run(ctx context.Context) {
+	wait.NonSlidingUntil(func() {
+		if err := desch.deschedulerOnce(ctx); err != nil {
+			klog.Errorf("Error descheduling pods: %v", err)
+		}
+	}, rs.DeschedulingInterval, ctx.Done())
+}
+```
+
+As mentioned earlier, only two extension points are available for executing. The remaining (PreSort, Filter, …) are expected to be invoked from within implementation of Deschedule resp. Balance extension points.
+
+### Descheduling profiles
+
+A profile is the smallest configuration unit that allows to run the same strategies with different configuration over different sets of pods (specified by combination of included/excluded namespaces and a label selector). The descheduling profile is not the same as the scheduling profile (a pod is not allowed to specify which descheduler strategy is to be run when the pod is considered for eviction). Thus, each descheduling profile is rather a representation of descheduling strategies sharing the same eviction policies (e.g. identical priority threshold, target node preferences).
+
+Plugins are expected to be executed in the same order as they are specified across and in each profile. As mentioned previously all `Deschedule` extension points are executed before any `Balance` extension point is (including cross-profile) to aim for even balancing when configured.
+
+
+
+### Descheduling workflow
+
+The following diagram depicts the descheduling cycle workflow. In each descheduling cycle a list of nodes is passed to each Deschedule and Balance plugin. Both extension points are expected to list all pods on each node before further processing. First, all Deschedule extension points from all profiles are executed. Followed by all Balance extension points. The framework can be configured to keep a track of pods evicted per node/namespaces per each descheduling cycle. When a limit on the number of evicted pods is configured and exceeded, the descheduling cycle is prematurely finished. So it’s recommended to put the profiles in a way that the most preferable pods are evicted first.
+
+![Descheduling framework workflow diagram](framework-workflow-diagram.png)
```

Comment:
Right now we execute all deschedule of all profiles, then all balance of all profiles. After some discussion that was the decision we made, right? Are we going back to executing everything from each profile before going to next?

Reply:
Oh, I see the inconsistency in the diagram. Now the diagram gives impression we execute more than the `Deschedule` extension point when going through all profiles. The goal was to make a diagram where we go through `Deschedule` extension point of all profiles (thus `Profile[0]` and `Profile[n]` followed by again `Profile[0]` and `Profile[n]`).

Reply:
Updating in https://github.com/kubernetes-sigs/descheduler/pull/1396